File: MoonSimulation/Renderers/MoonPhaseRenderer.cs
using SkiaSharp;

namespace MoonSimulation.Renderers;

/// <summary>
/// Shared moon phase shadow algorithm — the single source of truth
/// for the elliptical terminator technique used by both MoonPhaseView
/// and the mini moon in the sky scene.
/// </summary>
public static class MoonPhaseRenderer
{
    /// <summary>
    /// Draws the phase shadow overlay on a moon circle using the elliptical terminator technique.
    /// At moonAngleDeg 0° = New Moon (fully shadowed), 180° = Full Moon (fully lit).
    /// </summary>
    public static void DrawPhaseOverlay(SKCanvas canvas, SKPoint center, float radius,
        double moonAngleDeg, SKColor shadowColor)
    {
        double angle = ((moonAngleDeg % 360) + 360) % 360;
        double phaseRad = angle * Math.PI / 180.0;

        canvas.Save();

        // Clip to the moon circle
        using var clipPath = new SKPath();
        clipPath.AddCircle(center.X, center.Y, radius);
        canvas.ClipPath(clipPath);

        using var shadowPaint = new SKPaint
        {
            IsAntialias = true,
            Color = shadowColor,
            Style = SKPaintStyle.Fill
        };

        float terminatorScale = (float)Math.Cos(phaseRad);

        using var shadowPath = new SKPath();

        if (angle <= 180)
        {
            // Waxing phases: right side illuminated, shadow on the left
            shadowPath.MoveTo(center.X, center.Y - radius);

            // Left arc (left half of moon)
            shadowPath.ArcTo(
                new SKRect(center.X - radius, center.Y - radius,
                           center.X + radius, center.Y + radius),
                -90, -180, false);

            // Terminator arc (elliptical)
            float termW = radius * Math.Abs(terminatorScale);
            shadowPath.ArcTo(
                new SKRect(center.X - termW, center.Y - radius,
                           center.X + termW, center.Y + radius),
                90, terminatorScale >= 0 ? -180 : 180, false);

            shadowPath.Close();
        }
        else
        {
            // Waning phases: left side illuminated, shadow on the right
            shadowPath.MoveTo(center.X, center.Y - radius);

            // Right arc (right half of moon)
            shadowPath.ArcTo(
                new SKRect(center.X - radius, center.Y - radius,
                           center.X + radius, center.Y + radius),
                -90, 180, false);

            // Terminator arc (elliptical)
            float termW = radius * Math.Abs(terminatorScale);
            shadowPath.ArcTo(
                new SKRect(center.X - termW, center.Y - radius,
                           center.X + termW, center.Y + radius),
                90, terminatorScale >= 0 ? 180 : -180, false);

            shadowPath.Close();
        }

        canvas.DrawPath(shadowPath, shadowPaint);

        canvas.Restore();
    }
}

File: web/js/renderers/moonPhaseRenderer.js
// Shared moon phase shadow algorithm — the single source of truth
// for the elliptical terminator technique.

export function drawPhaseOverlay(ctx, cx, cy, radius, moonAngleDeg, shadowColor) {
  let angle = ((moonAngleDeg % 360) + 360) % 360;
  const phaseRad = (angle * Math.PI) / 180.0;

  ctx.save();

  // Clip to the moon circle
  ctx.beginPath();
  ctx.arc(cx, cy, radius, 0, Math.PI * 2);
  ctx.clip();

  ctx.fillStyle = shadowColor;

  const terminatorScale = Math.cos(phaseRad);
  const termW = radius * Math.abs(terminatorScale);

  ctx.beginPath();

  if (angle <= 180) {
    // Waxing phases: right side illuminated, shadow on the left
    // Start at top
    ctx.moveTo(cx, cy - radius);

    // Left arc (left half of moon) — sweep from -90° by -180° (clockwise)
    ctx.arc(cx, cy, radius, -Math.PI / 2, Math.PI / 2, true);

    // Terminator arc (elliptical) from bottom to top
    if (terminatorScale >= 0) {
      // Terminator bulges right — sweep counterclockwise
      ctx.ellipse(cx, cy, termW, radius, 0, Math.PI / 2, -Math.PI / 2, true);
    } else {
      // Terminator bulges left — sweep clockwise
      ctx.ellipse(cx, cy, termW, radius, 0, Math.PI / 2, -Math.PI / 2, false);
    }
  } else {
    // Waning phases: left side illuminated, shadow on the right
    // Start at top
    ctx.moveTo(cx, cy - radius);

    // Right arc (right half of moon) — sweep from -90° by +180° (counterclockwise)
    ctx.arc(cx, cy, radius, -Math.PI / 2, Math.PI / 2, false);

    // Terminator arc (elliptical) from bottom to top
    if (terminatorScale >= 0) {
      // Sweep clockwise
      ctx.ellipse(cx, cy, termW, radius, 0, Math.PI / 2, -Math.PI / 2, false);
    } else {
      // Sweep counterclockwise
      ctx.ellipse(cx, cy, termW, radius, 0, Math.PI / 2, -Math.PI / 2, true);
    }
  }

  ctx.closePath();
  ctx.fill();

  ctx.restore();
}
